README
NAME

    Dist::Zilla::Plugin::Code - dynamically create plugins from a bundle

VERSION

    Version 0.007

DESCRIPTION

    Dist::Zilla::Plugin::Code::AfterBuild

      something that runs after building is mostly complete

    Dist::Zilla::Plugin::Code::AfterRelease

      something that runs after release is mostly complete

    Dist::Zilla::Plugin::Code::BeforeArchive

      something that runs before the archive file is built

    Dist::Zilla::Plugin::Code::BeforeBuild

      something that runs before building really begins

    Dist::Zilla::Plugin::Code::BeforeRelease

      something that runs before release really begins

    Dist::Zilla::Plugin::Code::BuildRunner

      something used as a delegating agent during 'dzil run'

    Dist::Zilla::Plugin::Code::EncodingProvider

      something that sets a files' encoding

    Dist::Zilla::Plugin::Code::FileFinder

      something that finds files within the distribution

    Dist::Zilla::Plugin::Code::FileGatherer

      something that gathers files into the distribution

    Dist::Zilla::Plugin::Code::FileMunger

      something that munges files within the distribution

    Dist::Zilla::Plugin::Code::FilePruner

      something that prunes files from the distribution

    Dist::Zilla::Plugin::Code::Initialization

      something that runs when plugins are initialized

    Dist::Zilla::Plugin::Code::InstallTool

      something that creates an install program for a dist

    Dist::Zilla::Plugin::Code::LicenseProvider

      something that offers a license for a dist

    Dist::Zilla::Plugin::Code::MetaProvider

      something that provides data to merge into the distribution metadata

    Dist::Zilla::Plugin::Code::NameProvider

      something that provides a name for the dist

    Dist::Zilla::Plugin::Code::PrereqSource

      something that registers prereqs of the dist

    Dist::Zilla::Plugin::Code::ReleaseStatusProvider

      something that provides a release status for the dist

    Dist::Zilla::Plugin::Code::Releaser

      something that makes a release of the dist

    Dist::Zilla::Plugin::Code::TestRunner

      something that tests the dist

    Dist::Zilla::Plugin::Code::VersionProvider

      something that provides a version number for the dist

    Dist::Zilla::PluginBundle::Code

      a dynamic bundle

SUPPORT

 Bugs / Feature Requests

    Please report any bugs or feature requests through the issue tracker at
    https://github.com/skirmess/Dist-Zilla-Plugin-Code/issues. You will be
    notified automatically of any progress on your issue.

 Source Code

    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    https://github.com/skirmess/Dist-Zilla-Plugin-Code

      git clone https://github.com/skirmess/Dist-Zilla-Plugin-Code.git

AUTHOR

    Sven Kirmess <sven.kirmess@kzone.ch>

SEE ALSO

    Dist::Zilla



Changes
Revision history for Dist-Zilla-Plugin-Code

{{$NEXT}}

0.008  2023-08-09 20:40:36 UTC

  - Fixed incorrect usage of use_ok in tests to work with Perl 5.39.1
  - Switch license to the ISC license

0.007  2021-08-26 22:01:03 UTC

  - The Code::FileMunger plugin supports the munge_files method

0.006  2021-07-25 00:12:47 UTC

  - Documentation update

0.005  2021-07-24 23:14:09 UTC

  - Documentation update

0.004  2021-07-24 20:46:01 UTC

  - Added a dependency on Config::MVP 2.200012 (https://github.com/rjbs/Config-MVP/issues/13)

0.003  2020-08-10 16:36:00 UTC

  - Fixed link in documentation

0.002  2020-08-07 16:19:52 UTC

  - New module Dist::Zilla::Plugin::Code::Initialization
  - New module Dist::Zilla::PluginBundle::Code
  - Fixed synopsis
  - Improved error message if required arg is not given

0.001  2020-08-01 23:54:13 UTC

  - Initial release


t/lib/Local/FileConsumer.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::FileConsumer;

our $VERSION = '0.001';

use Moose;
with(
    'Dist::Zilla::Role::AfterBuild',
    'Dist::Zilla::Role::FileFinderUser' => {
        default_finders => ['ThisPluginDoesNotExist'],
    },
);

our $RESULT;

use namespace::autoclean;

sub after_build {
    my ($self) = @_;

    for my $file ( @{ $self->found_files } ) {
        push @{$RESULT}, $file->name;
    }

    return;
}

__PACKAGE__->meta->make_immutable;

1;


t/lib/Local/PluginBundle.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::PluginBundle;

our $VERSION = '0.001';

use Moose;
with 'Dist::Zilla::Role::PluginBundle';

our $RESULT;

use namespace::autoclean;

sub bundle_config {
    my ( $class, $section ) = @_;

    my $plugin_class = $section->{payload}{plugin};
    my $plugin_name  = $section->{payload}{name};
    my $input        = $section->{payload}{input};
    my $code         = $section->{payload}{code};
    $RESULT = undef;

    return [
        $plugin_name,
        $plugin_class,
        {
            $code => sub {
                my ($self) = @_;
                $self->log($input);
                $RESULT = $input * $input;
                return;
            },
        },
    ];
}

__PACKAGE__->meta->make_immutable;

1;


t/lib/Local/PluginBundleEasy.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::PluginBundleEasy;

our $VERSION = '0.001';

use Moose;
with 'Dist::Zilla::Role::PluginBundle::Easy';

our $RESULT;

use namespace::autoclean;

sub configure {
    my ($self) = @_;

    my $plugin_class = $self->payload->{plugin};
    my $plugin_name  = $self->payload->{name};
    my $input        = $self->payload->{input};
    my $code         = $self->payload->{code};
    $RESULT = undef;

    my ($moniker) = $plugin_class =~ m{ ^ \QDist::Zilla::Plugin::\E ( .+ ) }xsm;

    $self->add_plugins(
        [
            $moniker,
            $plugin_name,
            {
                $code => sub {
                    my ($self) = @_;
                    $self->log($input);
                    $RESULT = $input * $input;
                    return;
                },
            },
        ],
    );

    return;
}

__PACKAGE__->meta->make_immutable;

1;


Makefile.PL
use 5.006;
use strict;
use warnings;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
    ABSTRACT           => 'dynamically create plugins from a bundle',
    AUTHOR             => 'Sven Kirmess <sven.kirmess@kzone.ch>',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    DISTNAME         => 'Dist-Zilla-Plugin-Code',
    LICENSE          => 'open_source',
    MIN_PERL_VERSION => '5.006',
    NAME             => 'Dist::Zilla::Plugin::Code',
    PREREQ_PM        => {
        Carp                                       => 0,
        'Config::MVP'                              => '2.200012',
        'Dist::Zilla::Role::AfterBuild'            => 0,
        'Dist::Zilla::Role::AfterRelease'          => 0,
        'Dist::Zilla::Role::BeforeArchive'         => 0,
        'Dist::Zilla::Role::BeforeBuild'           => 0,
        'Dist::Zilla::Role::BeforeRelease'         => 0,
        'Dist::Zilla::Role::BuildRunner'           => 0,
        'Dist::Zilla::Role::EncodingProvider'      => 0,
        'Dist::Zilla::Role::FileFinder'            => 0,
        'Dist::Zilla::Role::FileGatherer'          => 0,
        'Dist::Zilla::Role::FileMunger'            => 0,
        'Dist::Zilla::Role::FilePruner'            => 0,
        'Dist::Zilla::Role::InstallTool'           => 0,
        'Dist::Zilla::Role::LicenseProvider'       => 0,
        'Dist::Zilla::Role::MetaProvider'          => 0,
        'Dist::Zilla::Role::NameProvider'          => 0,
        'Dist::Zilla::Role::Plugin'                => 0,
        'Dist::Zilla::Role::PluginBundle'          => 0,
        'Dist::Zilla::Role::PrereqSource'          => 0,
        'Dist::Zilla::Role::ReleaseStatusProvider' => 0,
        'Dist::Zilla::Role::Releaser'              => 0,
        'Dist::Zilla::Role::TestRunner'            => 0,
        'Dist::Zilla::Role::VersionProvider'       => 0,
        Moose                                      => 0,
        'MooseX::Types::Moose'                     => 0,
        'namespace::autoclean'                     => 0,
        strict                                     => 0,
        warnings                                   => 0,
    },
    TEST_REQUIRES => {
        'Capture::Tiny'                         => 0,
        Cwd                                     => 0,
        'Dist::Zilla::File::InMemory'           => 0,
        'Dist::Zilla::Role::FileFinderUser'     => 0,
        'Dist::Zilla::Role::PluginBundle::Easy' => 0,
        Exporter                                => 0,
        'File::Basename'                        => 0,
        'File::Path'                            => '2.07',
        'File::Spec'                            => 0,
        'File::Spec::Functions'                 => 0,
        'Test::DZil'                            => 0,
        'Test::Fatal'                           => 0,
        'Test::More'                            => '0.88',
        lib                                     => 0,
    },
    VERSION => '0.007',
    test    => {
        TESTS => 't/*.t',
    },
);

my %FallbackPrereqs = (
    'Capture::Tiny'                            => 0,
    Carp                                       => 0,
    'Config::MVP'                              => '2.200012',
    Cwd                                        => 0,
    'Dist::Zilla::File::InMemory'              => 0,
    'Dist::Zilla::Role::AfterBuild'            => 0,
    'Dist::Zilla::Role::AfterRelease'          => 0,
    'Dist::Zilla::Role::BeforeArchive'         => 0,
    'Dist::Zilla::Role::BeforeBuild'           => 0,
    'Dist::Zilla::Role::BeforeRelease'         => 0,
    'Dist::Zilla::Role::BuildRunner'           => 0,
    'Dist::Zilla::Role::EncodingProvider'      => 0,
    'Dist::Zilla::Role::FileFinder'            => 0,
    'Dist::Zilla::Role::FileFinderUser'        => 0,
    'Dist::Zilla::Role::FileGatherer'          => 0,
    'Dist::Zilla::Role::FileMunger'            => 0,
    'Dist::Zilla::Role::FilePruner'            => 0,
    'Dist::Zilla::Role::InstallTool'           => 0,
    'Dist::Zilla::Role::LicenseProvider'       => 0,
    'Dist::Zilla::Role::MetaProvider'          => 0,
    'Dist::Zilla::Role::NameProvider'          => 0,
    'Dist::Zilla::Role::Plugin'                => 0,
    'Dist::Zilla::Role::PluginBundle'          => 0,
    'Dist::Zilla::Role::PluginBundle::Easy'    => 0,
    'Dist::Zilla::Role::PrereqSource'          => 0,
    'Dist::Zilla::Role::ReleaseStatusProvider' => 0,
    'Dist::Zilla::Role::Releaser'              => 0,
    'Dist::Zilla::Role::TestRunner'            => 0,
    'Dist::Zilla::Role::VersionProvider'       => 0,
    Exporter                                   => 0,
    'File::Basename'                           => 0,
    'File::Path'                               => '2.07',
    'File::Spec'                               => 0,
    'File::Spec::Functions'                    => 0,
    Moose                                      => 0,
    'MooseX::Types::Moose'                     => 0,
    'Test::DZil'                               => 0,
    'Test::Fatal'                              => 0,
    'Test::More'                               => '0.88',
    lib                                        => 0,
    'namespace::autoclean'                     => 0,
    strict                                     => 0,
    warnings                                   => 0,
);

if ( !eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

if ( !eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

WriteMakefile(%WriteMakefileArgs);

# vim: ts=4 sts=4 sw=4 et: syntax=perl
